### DISCUSSION TITLE
Support for simple memory allocators

### DISCUSSION THREAD

**Discussion by leonsal** [stclib/STC#71] [Ideas]
Hi,

It seems STC interface to memory allocators is composed of the functions:
`malloc(), calloc(), realloc()` and `free()`.
I saw that the vector implementation uses `realloc()`.

The problem with `realloc()` is that it could be difficult to
implement in simple memory allocators such as bump, pool or arena allocators,
because the size of the allocated area is not stored by the allocator.
So its `realloc()` will not be capable of `memcpy()` the previous area to the new one.

One possible solution is to STC to use only `malloc()` and `free()`
and implement reallocation by itself.
C++ [memory resources](https://en.cppreference.com/w/cpp/memory/memory_resource) abstract interface only uses:
`do_allocate()` and `do_deallocate()`.

Other alternative is a weaker form of `realloc()` where the size of
previous area is supplied:
`void* realloc2(void* p, size_t prev_size, size_t new_size);`

I think that support for these simple memory allocators are important for
real-time applications and games.
They also could simplify memory management for any application
if the lifetime of the allocations can be grouped.

Please, see:
[Untangling Lifetimes: The Arena Allocator](https://www.rfleury.com/p/untangling-lifetimes-the-arena-allocator)
[Arena allocator tips and tricks](https://nullprogram.com/blog/2023/09/27)
[Writing custom memory allocators in C++](https://thealexcons.github.io/custom-memory-allocators/index.html)

Also, it would be great to include in the Algorithms section
implementation of common simple memory allocators.

What are your thoughts ?

Regards.

**Comment by tylov** [comment]
Hi, thanks for your suggestions. This it is one thing I have considered for a while. Thanks for reminding. Basically all heap memory allocated by STC is already stored in the containers which allocate, so it should be fairly straight forward to upgrade the API for this, but still a bit work. Also it means some API changes, although API breaks are accepable when there are good reasons for it. I will likely go with malloc(sz) / realloc2(p, prev_sz, sz) / free2(p, sz) variants with sizes, because I still prefer to use realloc as default due to performance. realloc2() can be implemented by free2() + malloc() anyway.

**Comment by tylov** [comment]
Added old-size parameters for i_realloc() and size for i_free(). This will break some code if they use custom allocators, but is reasonable, as it should affect very little code in most cases, and is a trivial change. 
Note: I have not yet made a custom allocator that counts malloc/realloc vs free to check that it sums to zero at program end.

**Comment by leonsal** [comment]
I coded a simple [test](https://github.com/leonsal/stc_tests) for `cvec` and `cmap` using custom allocator.
The number of allocations/deallocations sums to zero for `cvec` but not for `cmap`.

**Comment by tylov** [comment]
Should be fixed in [in this commit](https://github.com/stclib/STC/commit/85f6fc2855f4384b29858e9c2a3ba60122631555?diff=unified). I have tested with the demos.c example which uses most of the containers.

PS. If you use a custom allocator that uses a global context, you can define STC_ALLOCATOR instead of i_allocator. Include my_alloc.h on top of your file and it will be used by all containers implemented in the TU.

**Comment by tylov** [comment]
You forgot to multiply n*sz in calloc, so it works fine now. Reopen this if you find other issues.
